### ISSUE TITLE
Jolie installer broken on Windows

### ISSUE THREAD

**Issue by mwallnoefer** [jolie/jolie#220]
Just for interest I have tried to run the Jolie testsuite on Windows to check if nothing had been broken, but unfortunately it fails already on an include:
```
PS C:\Jolie\test> jolie --charset "utf-8" --trace test.ol
Illegal char <:> at index 4: file:/C:/Jolie/include/types/JavaException.iol
```

Don't know if we still should pay attention to Windows or let it go...
We often had breakages, some of them I have fixed last year.

**Comment by fmontesi** [comment]
We have Windows users (e.g., here on github @bmaschio and @mmontesi), so it's important.

The problem here is that we lack an automated testing pipeline for Windows. Anybody with experience/willingness to set one up?

**Comment by mmontesi** [comment]
I confirm issues on Windows.
I can't run the installer itself (same error reported by @mwallnoefer ):
```
PS C:\Users\ma.montesi\Downloads> java -jar .\jolie-1.9.0.jar
Illegal char <:> at index 4: file:/C:/Users/MA4509~1.MON/AppData/Local/Temp/jolie_installer_tmp2249035435699964779/dist/
jolie/include/types/JavaException.iol
```

**Comment by fmontesi** [comment]
I've set up GitHub Actions:

https://github.com/jolie/jolie/blob/master/.github/workflows/maven.yml

But for now it tests only on Linux, as you can see. Github actions supports matrices now, for testing in parallel on multiple java versions and oses, and this includes windows and macos.

Somebody willing to have a look at this? It shouldn't be super complex, and it's really really important. @bmaschio, @mwallnoefer, @mmontesi?

**Comment by mwallnoefer** [comment]
I have done a git-bisect. The commit which introduced the breakage are 5565c7c174872eb719e7e8ef05b80ac2176a3891 or bc7c8a4082fd0e829c3ce8a0db6e5fd3cc213238.

**Comment by mwallnoefer** [comment]
First attempt 3126229645534de0d74c034dc268d7b0a575b08c which makes the illegal char error disappear and tries also the other handlers, but it is still not enough.

**Comment by mwallnoefer** [comment]
@fmontesi  I guess that we need to change the method _guessIncludeFilepath()_ to not create an URI with `new URI(path)` [the Unix way] but with `new File(path).toURI()` which is portable. Most probably this has to be done in line 690 but I am not completely sure and I don't want to introduce another breakage.

Btw. we may discuss this also on Slack.

**Comment by fmontesi** [comment]
I've pushed a fix.
Please try this installer:
[jolie_installer.zip](https://github.com/jolie/jolie/files/4544391/jolie_installer.zip)
(You have to unzip the file and then run the jar.)

**Comment by mwallnoefer** [comment]
Okay, the situation became better, but we are still failing some tests, e.g. the JAP one: 
[jolie.log](https://github.com/jolie/jolie/files/4544905/jolie.log)

I have attempted to catch some more `Paths` exceptions, but the behaviour does not change: 72e9351dcf115a18e2591139aece09f091289559

**Comment by fmontesi** [comment]
Thanks for those patches!
Yeah, I haven't had time to debug the failing jap test. It's a new feature of 1.9.0 and it's still undocumented, while the fix to the installer was urgent, so I've pushed it.
I'll open a separate issue for that test. I'm closing this for now, assuming that the installer works for you (right?). :-)

**Comment by fmontesi** [comment]
Whoops, apparently there was still a little bug left. Here's the latest version.
[installer.zip](https://github.com/jolie/jolie/files/4545682/installer.zip)

**Comment by fmontesi** [comment]
@thesave @mwallnoefer @mmontesi 

Now it doesn't work, but because of windows_installer.ol in release_tools. :o)

**Comment by fmontesi** [comment]
@thesave has pushed a fix, thanks @thesave :-)

Here's the installer, for testing.
[jolie_installer.zip](https://github.com/jolie/jolie/files/4556497/jolie_installer.zip)

**Comment by mmontesi** [comment]
Last installer seems to solve all Windows issues.
Good job @thesave

**Comment by unibogdv** [comment]
I'm still getting the following error (I'm running the installer from an elevated command prompt):
```
Directory C:\Jolie does not exist. It has now been created.
apr 30, 2020 1:35:41 PM windows_installer.ol
INFO: [windows_installer.ol] Thrown unhandled fault: CannotCopyBins
Content (if any)
        .message[0] = Error: the installer could not find the launchers files. To complete the installation, try to run the installer with administrator privileges. : string

Error: the installer could not find the launchers files. To complete the installation, try to run the installer with administrator privileges.
```

**Comment by fmontesi** [comment]
Thanks @mmontesi. I'm closing this for now, please reopen it if it happens again.

**Comment by unibogdv** [comment]
@fmontesi the fixed installer for Windows is not yet in the Download section of the Jolie website.
The link "https://www.jolie-lang.org/downloads.html" is still pointing to the bugged installer.

**Comment by mwallnoefer** [comment]
Unfortunately I have seen the Jolie installer on Windows (Oracle JRE 8) to break again between Jolie _1.10 Beta 2_ and _Beta 3_:
```
PS C:\Users\Matthias\Downloads> java -jar jolie-1.10-beta2.jar

Insert the path for the environment variable JOLIE_HOME.
JOLIE_HOME indicates the directory in which the Jolie libraries will be installed.
[press Enter to use the default value: C:\Jolie]
Please note that using spaces in paths may cause problems.

 >
```
```
PS C:\Users\Matthias\Downloads> java -jar jolie-1.10-beta3.jar
Error: Unable to access jarfile jolie-1.10-beta3.jar
```
```
PS C:\Users\Matthias\Downloads> java -jar jolie-1.10.5.jar
Error: A J
```
Please be careful when touching the path escape code, we had so many breakages in the past.

**Comment by mwallnoefer** [comment]
Always regarding _Windows_ to not forget: please also pay attention to the the still unfixed test breakage introduced in commit https://github.com/jolie/jolie/commit/9edec818271fd11138ca56e351ad5421f3022f4b.

**Comment by fmontesi** [comment]
I've pushed a tentative fix, can you test?

**Comment by mwallnoefer** [comment]
Hi Fabrizio,

nope, the commit https://github.com/jolie/jolie/commit/bc2ab24978380bea5b908e248fbd4e98b17bef99 doesn't solve the issue. Furthermore it raises the compatibility level to Java 11, but haven't we agreed to continue to support Java 8?

The real fix is to change `req = "jolie"` (in the line 8 of both the test cases) into `req = "jolie.bat"` on Windows. But how to do that on the fly?

**Comment by fmontesi** [comment]
Ah, thanks for the info, that's enlightening!

Is there a way of having "jolie" make sense in Windows? For example, what
if we install a "jolie.cmd" in addition to "jolie.bat" (same content)? I
suspect that a jolie.cmd file can be launched as jolie from the console,
differently from a bat file, but I don't actually know.

If that fails, we can either:
- Add a java service operation that returns the OS that Jolie is running on.
- Add an (optional) parameter to ***@***.*** that automatically injects
".bat" to the executable name on Windows.

**Comment by mwallnoefer** [comment]
> Furthermore it raises the compatibility level to Java 11, but haven't we agreed to continue to support Java 8?

Ok, yes. Shouldn't we revert the commit first?

> - Add a java service operation that returns the OS that Jolie is running on.

This sounds feasible. Maybe just check for the _windir_ env variable (reference: https://www.dowdandassociates.com/blog/content/howto-set-an-environment-variable-in-windows-command-line-and-registry/)?

**Comment by fmontesi** [comment]
Ad Java 11: Ah, I forgot to reply about this point. We actually bumped the required version of Java to 11 some time ago (some other modules require it already.) The download page states this correctly: https://www.jolie-lang.org/downloads.html (Jolie 1.10.0 might work with Java 8).

Ad detecting the operating system: I think that we have OSUtils that we can use (we implemented a utility library for this a while ago).

Ad this issue: After talking to @mmontesi, I think that even putting `jolie.bat` won't be enough. We'll have to run "cmd /s" on windows. This is a common issue: we want to launch the program in a shell here (otherwise the bat won't be interpreted correctly).
New proposal: add a `shell?:bool` parameter to `exec` (default: false, for backwards compat). If set to `true`, then we launch the command under a shell, e.g., `cmd /s command` or `sh -s command`. What about that?

**Comment by mwallnoefer** [comment]
My proposal which works: https://github.com/jolie/jolie/tree/patch-windir

**Comment by fmontesi** [comment]
"It works", the two magical words that I have a soft side for! :-D Can you make a PR with that and same fix for the other 2 tests (they have similar code)? We can improve how to handle this later. Making the tests work definitely takes priority over beauty.

**Comment by mwallnoefer** [comment]
PR #353

**Comment by mwallnoefer** [comment]
> We actually bumped the required version of Java to 11 some time ago (some other modules require it already.) The download page states this correctly: https://www.jolie-lang.org/downloads.html

After upgrading to Java 11 the installer is working correctly also on Windows. Wow 😆 .

**Comment by fmontesi** [comment]
Woohoo! :+1:
